assertj-core/src/main/java/org/assertj/core/api/AssertFactory.java

```diff
@@ -29,12 +33,28 @@
 public interface AssertFactory<T, ASSERT extends Assert<?, ?>> {
 
   /**
-   * Creates the custom {@link Assert} instance for the given element value.
+   * Create the custom {@link Assert} instance for the given element value.
    * <p>
    * Typically, this will just invoke <code>assertThat(actual)</code>
    * @param actual the input value for the {@code Assert} instance
    * @return returns the custom {@code Assert} instance for the given element value
    */
   ASSERT createAssert(T actual);
 
+  /**
+   * Return the {@link Type} of the input this factory expects, or an empty {@code Optional}
+   * if the information is not available.
+   * 
+   * @implSpec Overriding implementations providing a non-empty optional are required to
+   * return a {@code Type} instance consistent with the value of the factory type parameter {@link T}.
```

Comment:
The need to explain this highlights the fact that the solution is suboptimal. I'll experiment with the `createAssert(Function)` alternative.